File: frontend/lib/models/playlist.dart
/// Playlist models for Astro.FM.
/// Represents songs and playlist results from the backend API.

class Song {
  final String id;
  final String title;
  final String artist;
  final String album;
  final int durationSeconds;
  final int bpm;
  final int energy;
  final int valence;
  final int danceability;
  final int acousticness;
  final int instrumentalness;
  final List<String> genres;
  final List<String> moods;
  final List<String> elements;
  final List<String> planetaryEnergy;
  final int intensity;
  final String? modality;
  final List<String>? timeOfDay;

  Song({
    required this.id,
    required this.title,
    required this.artist,
    required this.album,
    required this.durationSeconds,
    required this.bpm,
    required this.energy,
    required this.valence,
    required this.danceability,
    required this.acousticness,
    required this.instrumentalness,
    required this.genres,
    required this.moods,
    required this.elements,
    required this.planetaryEnergy,
    required this.intensity,
    this.modality,
    this.timeOfDay,
  });

  factory Song.fromJson(Map<String, dynamic> json) {
    return Song(
      id: json['id'] as String,
      title: json['title'] as String,
      artist: json['artist'] as String,
      album: json['album'] as String,
      durationSeconds: json['duration_seconds'] as int,
      bpm: json['bpm'] as int,
      energy: json['energy'] as int,
      valence: json['valence'] as int,
      danceability: json['danceability'] as int,
      acousticness: json['acousticness'] as int,
      instrumentalness: json['instrumentalness'] as int,
      genres: List<String>.from(json['genres'] as List),
      moods: List<String>.from(json['moods'] as List),
      elements: List<String>.from(json['elements'] as List),
      planetaryEnergy: List<String>.from(json['planetary_energy'] as List),
      intensity: json['intensity'] as int,
      modality: json['modality'] as String?,
      timeOfDay: json['time_of_day'] != null 
          ? List<String>.from(json['time_of_day'] as List)
          : null,
    );
  }

  /// Get formatted duration string (MM:SS)
  String get formattedDuration {
    final minutes = durationSeconds ~/ 60;
    final seconds = durationSeconds % 60;
    return '$minutes:${seconds.toString().padLeft(2, '0')}';
  }
}

class PlaylistResult {
  final List<Song> songs;
  final int totalDurationSeconds;
  final double vibeMatchScore;
  final List<int> energyArc;
  final Map<String, int> elementDistribution;
  final Map<String, int> moodDistribution;
  final Map<String, dynamic> generationMetadata;
  final double durationMinutes;
  final int songCount;

  PlaylistResult({
    required this.songs,
    required this.totalDurationSeconds,
    required this.vibeMatchScore,
    required this.energyArc,
    required this.elementDistribution,
    required this.moodDistribution,
    required this.generationMetadata,
    required this.durationMinutes,
    required this.songCount,
  });

  factory PlaylistResult.fromJson(Map<String, dynamic> json) {
    return PlaylistResult(
      songs: (json['songs'] as List)
          .map((songJson) => Song.fromJson(songJson as Map<String, dynamic>))
          .toList(),
      totalDurationSeconds: json['total_duration_seconds'] as int,
      vibeMatchScore: (json['vibe_match_score'] as num).toDouble(),
      energyArc: List<int>.from(json['energy_arc'] as List),
      elementDistribution: Map<String, int>.from(
        (json['element_distribution'] as Map).map(
          (key, value) => MapEntry(key as String, value as int),
        ),
      ),
      moodDistribution: Map<String, int>.from(
        (json['mood_distribution'] as Map).map(
          (key, value) => MapEntry(key as String, value as int),
        ),
      ),
      generationMetadata: json['generation_metadata'] as Map<String, dynamic>,
      durationMinutes: (json['duration_minutes'] as num).toDouble(),
      songCount: json['song_count'] as int,
    );
  }

  /// Get formatted total duration (e.g., "1h 10m" or "45m")
  String get formattedDuration {
    final hours = (durationMinutes / 60).floor();
    final minutes = (durationMinutes % 60).round();
    
    if (hours > 0) {
      return '${hours}h ${minutes}m';
    }
    return '${minutes}m';
  }
}

/// Track from the 114K music dataset.
class DatasetTrack {
  final String trackId;
  final String trackName;
  final String artists;
  final String albumName;
  final int durationMs;
  final int popularity;
  final double energy;
  final double valence;
  final double danceability;
  final String? mainGenre;
  final String? subgenre;
  final String? element;

  DatasetTrack({
    required this.trackId,
    required this.trackName,
    required this.artists,
    required this.albumName,
    required this.durationMs,
    required this.popularity,
    required this.energy,
    required this.valence,
    required this.danceability,
    this.mainGenre,
    this.subgenre,
    this.element,
  });

  factory DatasetTrack.fromJson(Map<String, dynamic> json) {
    return DatasetTrack(
      trackId: json['track_id'] as String,
      trackName: json['track_name'] as String,
      artists: json['artists'] as String,
      albumName: json['album_name'] as String,
      durationMs: json['duration_ms'] as int,
      popularity: json['popularity'] as int,
      energy: (json['energy'] as num).toDouble(),
      valence: (json['valence'] as num).toDouble(),
      danceability: (json['danceability'] as num).toDouble(),
      mainGenre: json['main_genre'] as String?,
      subgenre: json['subgenre'] as String?,
      element: json['element'] as String?,
    );
  }

  /// Get formatted duration string (MM:SS)
  String get formattedDuration {
    final minutes = durationMs ~/ 60000;
    final seconds = (durationMs % 60000) ~/ 1000;
    return '$minutes:${seconds.toString().padLeft(2, '0')}';
  }

  /// Serialize to JSON for storage.
  Map<String, dynamic> toJson() => {
    'track_id': trackId,
    'track_name': trackName,
    'artists': artists,
    'album_name': albumName,
    'duration_ms': durationMs,
    'popularity': popularity,
    'energy': energy,
    'valence': valence,
    'danceability': danceability,
    'main_genre': mainGenre,
    'subgenre': subgenre,
    'element': element,
  };
}

/// Playlist result from the 114K music dataset.
class DatasetPlaylistResult {
  final List<DatasetTrack> tracks;
  final int totalDurationMs;
  final double vibeMatchScore;
  final List<double> energyArc;
  final Map<String, int> elementDistribution;
  final Map<String, int> genreDistribution;
  final Map<String, dynamic> generationMetadata;

  DatasetPlaylistResult({
    required this.tracks,
    required this.totalDurationMs,
    required this.vibeMatchScore,
    required this.energyArc,
    required this.elementDistribution,
    required this.genreDistribution,
    required this.generationMetadata,
  });

  factory DatasetPlaylistResult.fromJson(Map<String, dynamic> json) {
    return DatasetPlaylistResult(
      tracks: (json['tracks'] as List)
          .map((t) => DatasetTrack.fromJson(t as Map<String, dynamic>))
          .toList(),
      totalDurationMs: json['total_duration_ms'] as int,
      vibeMatchScore: (json['vibe_match_score'] as num).toDouble(),
      energyArc: (json['energy_arc'] as List).map((e) => (e as num).toDouble()).toList(),
      elementDistribution: Map<String, int>.from(
        (json['element_distribution'] as Map).map(
          (key, value) => MapEntry(key as String, value as int),
        ),
      ),
      genreDistribution: Map<String, int>.from(
        (json['genre_distribution'] as Map).map(
          (key, value) => MapEntry(key as String, value as int),
        ),
      ),
      generationMetadata: json['generation_metadata'] as Map<String, dynamic>,
    );
  }

  /// Serialize to JSON for storage.
  Map<String, dynamic> toJson() => {
    'tracks': tracks.map((t) => t.toJson()).toList(),
    'total_duration_ms': totalDurationMs,
    'vibe_match_score': vibeMatchScore,
    'energy_arc': energyArc,
    'element_distribution': elementDistribution,
    'genre_distribution': genreDistribution,
    'generation_metadata': generationMetadata,
  };

  /// Get formatted total duration (e.g., "1h 10m" or "45m")
  String get formattedDuration {
    final totalMinutes = totalDurationMs ~/ 60000;
    final hours = totalMinutes ~/ 60;
    final minutes = totalMinutes % 60;
    
    if (hours > 0) {
      return '${hours}h ${minutes}m';
    }
    return '${minutes}m';
  }

  int get trackCount => tracks.length;
}

File: backend/api/routes/playlist.py
"""
Playlist generation API routes.
"""
from datetime import datetime
from zoneinfo import ZoneInfo
from fastapi import APIRouter, HTTPException

from models.playlist import PlaylistRequest, PlaylistResult
from models.track_model import GenrePreference
from services.ephemeris import calculate_natal_chart
from services.vibe_calculator import calculate_vibe_parameters
from services.playlist_matcher import generate_playlist, generate_blended_playlist

router = APIRouter(tags=["playlist"])


@router.post("/generate", response_model=PlaylistResult)
async def generate_playlist_endpoint(request: PlaylistRequest) -> PlaylistResult:
    """
    Generate a personalized playlist based on birth chart and current transits.
    
    Args:
        request: Playlist generation request with birth data and preferences
        
    Returns:
        PlaylistResult with ordered songs, energy arc, and metadata
        
    Raises:
        HTTPException: 400 for invalid input, 500 for calculation errors
    """
    try:
        # Parse birth datetime
        try:
            birth_dt = datetime.fromisoformat(request.birth_datetime.replace('Z', '+00:00'))
        except ValueError as e:
            raise HTTPException(
                status_code=400,
                detail=f"Invalid birth_datetime format: {str(e)}. Use ISO format (YYYY-MM-DDTHH:MM:SS)"
            )
        
        # Convert to UTC if timezone is provided
        if request.timezone != "UTC":
            try:
                local_tz = ZoneInfo(request.timezone)
                birth_dt = birth_dt.replace(tzinfo=local_tz)
                birth_dt = birth_dt.astimezone(ZoneInfo("UTC"))
                birth_dt = birth_dt.replace(tzinfo=None)  # Remove tzinfo for calculation
            except Exception as e:
                raise HTTPException(
                    status_code=400,
                    detail=f"Invalid timezone: {request.timezone}"
                )
        
        # Parse current datetime (defaults to now if not provided)
        if request.current_datetime:
            try:
                current_dt = datetime.fromisoformat(request.current_datetime.replace('Z', '+00:00'))
                if request.timezone != "UTC":
                    current_dt = current_dt.replace(tzinfo=local_tz)
                    current_dt = current_dt.astimezone(ZoneInfo("UTC"))
                    current_dt = current_dt.replace(tzinfo=None)
            except ValueError as e:
                raise HTTPException(
                    status_code=400,
                    detail=f"Invalid current_datetime format: {str(e)}"
                )
        else:
            current_dt = datetime.utcnow()
        
        # Use current location if provided, otherwise use birth location
        current_lat = request.current_latitude if request.current_latitude is not None else request.latitude
        current_lon = request.current_longitude if request.current_longitude is not None else request.longitude
        
        # Step 1: Calculate natal chart
        try:
            natal_chart = calculate_natal_chart(
                birth_datetime=birth_dt,
                latitude=request.latitude,
                longitude=request.longitude
            )
        except Exception as e:
            raise HTTPException(
                status_code=500,
                detail=f"Error calculating natal chart: {str(e)}"
            )
        
        # Step 2: Calculate vibe parameters from chart + transits
        try:
            vibe_params = calculate_vibe_parameters(
                natal_chart=natal_chart,
                current_datetime=current_dt,
                latitude=current_lat,
                longitude=current_lon,
                genre_preferences=request.genre_preferences
            )
        except Exception as e:
            raise HTTPException(
                status_code=500,
                detail=f"Error calculating vibe parameters: {str(e)}"
            )
        
        # Step 3: Generate playlist using the AI-driven CosmicPlaylistBuilder (SongToSpot method)
        from services.cosmic.playlist_builder import get_playlist_builder
        from services.transits import get_current_moon_sign
        from models.song import Song
        import uuid

        builder = get_playlist_builder()
        
        # Determine signs from chart
        sun_sign = next((p["sign"] for p in natal_chart["planets"] if p["name"] == "Sun"), "Aries")
        moon_sign = next((p["sign"] for p in natal_chart["planets"] if p["name"] == "Moon"), "Water")
        rising_sign = natal_chart.get("ascendant_sign", "Aries")
        current_moon_sign, _ = get_current_moon_sign()

        # Extract genre names
        genre_list = request.genre_preferences if request.genre_preferences else ["Pop", "Indie", "Electronic"]

        # Generate via AI pipeline
        cosmic_result = await builder.generate_playlist(
            sun_sign=sun_sign,
            moon_sign=moon_sign,
            rising_sign=rising_sign,
            current_moon_sign=current_moon_sign,
            genre_preferences=genre_list,
            target_tracks=request.playlist_size
        )

        if not cosmic_result.success:
            raise HTTPException(status_code=500, detail=cosmic_result.error or "AI Generation failed")

        # Map CosmicPlaylistResult tracks to Song objects for frontend compatibility
        from data.constants import GENRES, MOODS, ELEMENTS, PLANETS
        
        songs = []
        for i, t in enumerate(cosmic_result.tracks):
            # Ensure valid ID format song_XXX
            song_id = f"song_{str(i+1).zfill(3)}"
            
            # Sanitize genres (must be in GENRES list)
            valid_song_genres = [g for g in genre_list if g in GENRES]
            if not valid_song_genres:
                valid_song_genres = ["Electronic"] if "Electronic" in GENRES else [GENRES[0]]
            
            # Sanitize moods (must be in MOODS list)
            valid_song_moods = [m for m in vibe_params.mood_direction if m in MOODS]
            if len(valid_song_moods) < 2:
                # Add fillers from constants to satisfy min_length=2
                fillers = ["Dreamy", "Uplifting", "Atmospheric"]
                for f in fillers:
                    if f in MOODS and f not in valid_song_moods:
                        valid_song_moods.append(f)
                    if len(valid_song_moods) >= 2: break
            
            # Ensure intensity and energy are ints (averaging the target range)
            energy_val = sum(vibe_params.target_energy) // 2
            valence_val = sum(vibe_params.target_valence) // 2
            intensity_val = sum(vibe_params.intensity_range) // 2
            bpm_val = 110 + (energy_val - 50) // 2 # Simple proxy for BPM
            
            songs.append(Song(
                id=song_id,
                title=t["name"],
                artist=t["artist"],
                album="Cosmic Selection",
                duration_seconds=210,
                bpm=int(bpm_val),
                energy=int(energy_val),
                valence=int(valence_val),
                danceability=int(energy_val * 0.9), # Proxy
                acousticness=50,
                instrumentalness=20,
                genres=valid_song_genres[:3],
                moods=valid_song_moods[:4],
                elements=[cosmic_result.element],
                planetary_energy=vibe_params.active_planets[:3],
                intensity=intensity_val,
                spotify_id=t.get("url", "").split("/")[-1] if "/" in t.get("url", "") else "",
            ))

        # Build final response
        playlist = PlaylistResult(
            songs=songs,
            total_duration_seconds=len(songs) * 210,
            vibe_match_score=95.0,
            energy_arc=[s.energy for s in songs],
            element_distribution={cosmic_result.element: len(songs)},
            mood_distribution={m: 0 for m in MOODS}, # Initialize
            generation_metadata={
                "method": "SongToSpot (AI)",
                "vibe_summary": cosmic_result.vibe_summary,
                "playlist_url": cosmic_result.playlist_url,
                "app_account": "palazques@gmail.com"
            }
        )
        
        # Populate mood distribution
        for s in songs:
            for m in s.moods:
                playlist.mood_distribution[m] = playlist.mood_distribution.get(m, 0) + 1
        
        return playlist
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Unexpected error: {str(e)}"
        )


from pydantic import BaseModel, Field
from typing import Optional, List


class BlendedPlaylistRequest(BaseModel):
    """Request for blended playlist generation (user library + app dataset)."""
    birth_datetime: str = Field(..., description="Birth date/time in ISO format")
    latitude: float = Field(..., ge=-90, le=90)
    longitude: float = Field(..., ge=-180, le=180)
    timezone: str = Field(default="UTC")
    playlist_size: int = Field(default=20, ge=10, le=30)
    current_datetime: Optional[str] = None
    current_latitude: Optional[float] = None
    current_longitude: Optional[float] = None
    # Genre preferences
    main_genres: List[str] = Field(default=[])
    subgenres: List[str] = Field(default=[])
    include_related: bool = Field(default=True)


@router.post("/generate-blended")
async def generate_blended_playlist_endpoint(request: BlendedPlaylistRequest):
    """
    Generate a cosmic playlist using the AI-powered CosmicPlaylistBuilder.
    
    Uses AI to curate track suggestions based on the user's birth chart
    and current transits, then resolves them on Spotify.
    
    Note: This endpoint now uses the same cosmic builder as /generate
    for consistent AI-powered playlist generation.
    """
    try:
        # Parse birth datetime
        try:
            birth_dt = datetime.fromisoformat(request.birth_datetime.replace('Z', '+00:00'))
        except ValueError as e:
            raise HTTPException(
                status_code=400,
                detail=f"Invalid birth_datetime format: {str(e)}"
            )
        
        # Convert to UTC if timezone is provided
        if request.timezone != "UTC":
            try:
                local_tz = ZoneInfo(request.timezone)
                birth_dt = birth_dt.replace(tzinfo=local_tz)
                birth_dt = birth_dt.astimezone(ZoneInfo("UTC"))
                birth_dt = birth_dt.replace(tzinfo=None)
            except Exception as e:
                raise HTTPException(status_code=400, detail=f"Invalid timezone: {request.timezone}")
        
        # Calculate natal chart
        natal_chart = calculate_natal_chart(
            birth_datetime=birth_dt,
            latitude=request.latitude,
            longitude=request.longitude
        )
        
        # Extract signs from chart
        sun_sign = next((p["sign"] for p in natal_chart["planets"] if p["name"] == "Sun"), "Aries")
        moon_sign = next((p["sign"] for p in natal_chart["planets"] if p["name"] == "Moon"), "Cancer")
        rising_sign = natal_chart.get("ascendant_sign", "Aries")
        
        # Get current moon sign
        from services.transits import get_current_moon_sign
        current_moon_sign, _ = get_current_moon_sign()
        
        # Build genre list from preferences
        genre_list = request.main_genres + request.subgenres if (request.main_genres or request.subgenres) else ["Pop", "Indie", "Electronic"]
        
        # Generate via AI pipeline (CosmicPlaylistBuilder)
        from services.cosmic.playlist_builder import get_playlist_builder
        builder = get_playlist_builder()
        
        cosmic_result = await builder.generate_playlist(
            sun_sign=sun_sign,
            moon_sign=moon_sign,
            rising_sign=rising_sign,
            current_moon_sign=current_moon_sign,
            genre_preferences=genre_list,
            target_tracks=request.playlist_size
        )
        
        if not cosmic_result.success:
            raise HTTPException(status_code=500, detail=cosmic_result.error or "AI Generation failed")
        
        # Return cosmic result in blended format for backward compatibility
        return {
            "tracks": [
                {
                    "track_id": f"cosmic_{i}",
                    "track_name": t["name"],
                    "artists": t["artist"],
                    "album_name": "Cosmic Selection",
                    "duration_ms": 210000,
                    "popularity": 80,
                    "energy": 0.7,
                    "valence": 0.6,
                    "danceability": 0.65,
                    "main_genre": genre_list[0] if genre_list else None,
                    "subgenre": None,
                    "element": cosmic_result.element,
                    "source": "cosmic_ai",
                    "match_score": 95.0,
                    "spotify_url": t.get("url", ""),
                    "album_art": t.get("album_art", ""),
                }
                for i, t in enumerate(cosmic_result.tracks)
            ],
            "total_duration_ms": len(cosmic_result.tracks) * 210000,
            "vibe_match_score": 95.0,
            "energy_arc": [0.7] * len(cosmic_result.tracks),
            "element_distribution": {cosmic_result.element: len(cosmic_result.tracks)},
            "genre_distribution": {genre_list[0]: len(cosmic_result.tracks)} if genre_list else {},
            "generation_metadata": {
                "source": "cosmic_ai",
                "method": "SongToSpot (AI)",
                "user_library_tracks": 0,
                "dataset_tracks": 0,
                "playlist_size_requested": request.playlist_size,
                "tracks_selected": len(cosmic_result.tracks),
                "vibe_summary": cosmic_result.vibe_summary,
                "playlist_url": cosmic_result.playlist_url,
            },
        }
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error generating cosmic playlist: {str(e)}")

